### PR TITLE
Implement dependency injection system

### PR DESCRIPTION
**Pull request by lhorie**
[Rendered RFC](https://github.com/fusionjs/rfcs/blob/di/text/0000-dependency-injection.md)

### CODE REVIEW

**Comment by lhorie** [comment]
Another unresolved question revolves around the reliance on Map/Set and the impact on bundle size when supporting e.g. Preact

**Comment by lhorie** [comment]
Another question: how is this supposed to work under the proposed DI system?

```js
const {fetch, ignore} = CsrfProtection.of();
const UniversalEvents = app.plugin(UniversalEventsPlugin, {fetch});
```

File: text/0000-dependency-injection.md [hunk 1]
```diff
@@ -0,0 +1,141 @@
...
+};
+```
+
+This is messy because it involves hard-coding test-related logic into the app, and polluting the API of the main export.
+
+### String tokens
+
+In AngularJS, the DI API required registering services to stringly-typed names. This comes with the limitation that two services can't have the same human-readable "name" and that there's no source of truth cataloguing what names are used within the DI container.
+
+This RFC solves these issues by making names unique and not stringly-typed. In addition, using unique tokens makes it more natural to support userland dependency registration.
```
Comment (AlexMSmithCA) [inline text/0000-dependency-injection.md]: Call out that this also makes Flow type checking easier?  Not sure how hard we are hoping to sell Flow here.

File: text/0000-dependency-injection.md [hunk 2]
```diff
@@ -0,0 +1,141 @@
...
+
+### Factories instead of app.configure
+
+It's possible to inject configuration using the dependency-less plugin pattern:
```
Comment (AlexMSmithCA) [inline text/0000-dependency-injection.md]: I'd even go so far as to stress that injecting configuration is as simple as registering another dependency, to be consumed by the plugin.

File: text/0000-dependency-injection.md [hunk 3]
```diff
@@ -0,0 +1,141 @@
...
+app.register(SomeConfigToken, config);
+```
+
+We opt to provide a helper to reduce the amount of boilerplate in config files.
+
+---
+
+# Unresolved questions
+
+* Should `app.register` support token-less registration? Why not just `app.register(createToken(''), thing)`?
```
Comment (AlexMSmithCA) [inline text/0000-dependency-injection.md]: If we opt for `app.register(plugin: PluginType, token: TokenType)` then I think it flows nicely into "well the token is optional, if you don't want it used as a dependency" without this weird 'create an empty token' convention.

File: text/0000-dependency-injection.md [hunk 4]
```diff
@@ -0,0 +1,141 @@
...
+* Feasibility of codemods; how easy is it to migrate userland plugins?
+* Reducing verbosity (related to flow.js, etc)
+* Type safety when overriding plugins (i.e. disallowing coercion to union types)
+* why `createToken` and not just use `Symbol`?
```
Comment (AlexMSmithCA) [inline text/0000-dependency-injection.md]: Hm, how might that work with Flow types?

**Pull request merged**